File: src/antares_cli/agent/model_turn.py
"""Single-turn model execution for the Antares agent loop."""

from __future__ import annotations

import json
from collections.abc import Callable
from dataclasses import dataclass
from typing import Protocol

from antares_cli.agent.contracts import BuildAgentStateFn, ProgressCallback
from antares_cli.agent.model_adapter import ModelAdapter, ToolCallResult
from antares_cli.agent.model_turn_stream import ModelTurnStreamer
from antares_cli.agent.state import ModelSessionState, TurnResult
from antares_cli.agent.streaming import ParsedToolCall
from antares_cli.inference.backend import InferenceBackend
from antares_cli.output.finding import Finding, TrajectoryEntry

MAX_POST_BUDGET_SUBMISSION_ATTEMPTS = 3


@dataclass(slots=True)
class _ToolExecutionResult:
    executed_tool_calls: list[ToolCallResult]
    submitted: bool = False
    all_duplicates: bool = False


class SubmitToolCallFn(Protocol):
    def __call__(
        self,
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
        *,
        progress_callback: ProgressCallback | None = None,
    ) -> list[Finding]: ...


class ModelTurnRunner:
    """Runs one model turn and translates streamed output into agent state changes."""

    def __init__(
        self,
        *,
        inference_backend: InferenceBackend | None,
        adapter: ModelAdapter,
        is_submit_tool_call: Callable[[ParsedToolCall], bool],
        execute_tool_call: Callable[[ParsedToolCall, ModelSessionState], str],
        handle_submit_tool_call: SubmitToolCallFn,
        build_agent_state: BuildAgentStateFn,
    ) -> None:
        self.adapter = adapter
        self.is_submit_tool_call = is_submit_tool_call
        self.execute_tool_call = execute_tool_call
        self.handle_submit_tool_call = handle_submit_tool_call
        self.build_agent_state = build_agent_state
        self.streamer = ModelTurnStreamer(
            inference_backend=inference_backend,
            adapter=adapter,
            build_agent_state=build_agent_state,
        )

    def run_turn(
        self,
        state: ModelSessionState,
        *,
        progress_callback: ProgressCallback | None = None,
        iteration_index: int = 0,
        maximum_iterations: int,
    ) -> TurnResult:
        """Run one full model turn: stream, parse events, execute tools, update messages."""
        parsed_turn = self.streamer.parse_turn_with_retries(
            state,
            progress_callback=progress_callback,
            retry_parse_errors=not _is_terminal_budget_exhausted(state),
        )

        if parsed_turn.generation_error:
            state.generation_errors += 1
            state.reasoning_log.append(
                "Model generation error interrupted this run; results may be incomplete."
            )
            return TurnResult(done=True)

        if _is_terminal_budget_exhausted(state):
            tool_execution = self._execute_tool_calls(
                parsed_turn.tool_calls,
                state,
                progress_callback=progress_callback,
            )
            if tool_execution.submitted or state.done_signaled:
                return TurnResult(done=True)
            return self._handle_post_budget_submission_attempt(
                state,
                assistant_text=parsed_turn.assistant_text,
                executed_tool_calls=tool_execution.executed_tool_calls,
                progress_callback=progress_callback,
            )

        if not parsed_turn.tool_calls:
            return self._handle_no_tool_calls(
                state,
                assistant_text=parsed_turn.assistant_text,
            )

        state.consecutive_no_tool_turns = 0
        tool_execution = self._execute_tool_calls(
            parsed_turn.tool_calls,
            state,
            progress_callback=progress_callback,
        )
        if tool_execution.submitted:
            return TurnResult(done=True)

        if tool_execution.all_duplicates:
            return self._handle_duplicate_tool_turn(
                state,
                assistant_text=parsed_turn.assistant_text,
            )

        state.consecutive_duplicate_turns = 0
        return self._handle_tool_results(
            state,
            assistant_text=parsed_turn.assistant_text,
            executed_tool_calls=tool_execution.executed_tool_calls,
            iteration_index=iteration_index,
            maximum_iterations=maximum_iterations,
            progress_callback=progress_callback,
        )

    def _handle_no_tool_calls(
        self,
        state: ModelSessionState,
        *,
        assistant_text: str,
    ) -> TurnResult:
        remaining_text = self.adapter.clean_model_text(assistant_text)
        if remaining_text and not state.answer_text:
            state.answer_text = remaining_text
        retry_index = state.consecutive_no_tool_turns
        if retry_index < self.adapter.no_tool_retry_limit and not state.done_signaled:
            state.consecutive_no_tool_turns += 1
            state.messages.append({"role": "assistant", "content": assistant_text})
            _append_traced_message(
                state,
                self.adapter.format_no_tool_retry(iteration_index=retry_index),
            )
            return TurnResult(done=False)
        return TurnResult(done=True)

    def _execute_tool_calls(
        self,
        tool_calls: list[ParsedToolCall],
        state: ModelSessionState,
        *,
        progress_callback: ProgressCallback | None,
    ) -> _ToolExecutionResult:
        result = _ToolExecutionResult(executed_tool_calls=[], all_duplicates=True)
        for parsed_call in tool_calls:
            if self.is_submit_tool_call(parsed_call):
                self.handle_submit_tool_call(
                    parsed_call,
                    state,
                    progress_callback=progress_callback,
                )
                result.submitted = True
                return result
            if self._skip_duplicate_tool_call(parsed_call, state):
                continue
            result.all_duplicates = False
            result.executed_tool_calls.append(self._execute_tool_call(parsed_call, state))
        return result

    @staticmethod
    def _skip_duplicate_tool_call(
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
    ) -> bool:
        dedup_key = json.dumps(
            {"tool": parsed_call.tool_name, "args": parsed_call.arguments},
            sort_keys=True,
        )
        if dedup_key not in state.seen_tool_calls:
            state.seen_tool_calls.add(dedup_key)
            return False
        state.tool_call_count += 1
        state.failed_tool_calls_count += 1
        message = f"Skipped duplicate tool call: {parsed_call.tool_name}"
        state.reasoning_log.append(message)
        state.session_trace.record_tool_call(
            tool_name=parsed_call.tool_name,
            arguments=parsed_call.arguments,
        )
        state.session_trace.record_event(
            phase="tool_blocked",
            payload={"tool_name": parsed_call.tool_name, "reason": message},
        )
        state.session_trace.record_tool_result(
            tool_name=parsed_call.tool_name,
            result_summary=message,
            evidence_id="duplicate_tool_call",
        )
        return True

    def _execute_tool_call(
        self,
        parsed_call: ParsedToolCall,
        state: ModelSessionState,
    ) -> ToolCallResult:
        return ToolCallResult(
            tool_name=parsed_call.tool_name,
            tool_response=self.execute_tool_call(parsed_call, state),
        )

    def _handle_duplicate_tool_turn(
        self,
        state: ModelSessionState,
        *,
        assistant_text: str,
    ) -> TurnResult:
        state.consecutive_duplicate_turns += 1
        state.messages.append({"role": "assistant", "content": assistant_text})
        force_submit = state.consecutive_duplicate_turns >= 3
        _append_traced_message(
            state,
            self.adapter.format_duplicate_tool_retry(force_submit=force_submit),
        )
        if force_submit:
            state.reasoning_log.append(
                "Loop stuck: 3 consecutive duplicate turns - forcing submit."
            )
            state.trajectory.append(
                TrajectoryEntry(
                    entry_type="think",
                    content="Stuck loop detected - forcing transition to submission.",
                )
            )
        return TurnResult(done=False)

    def _handle_post_budget_submission_attempt(
        self,
        state: ModelSessionState,
        *,
        assistant_text: str,
        executed_tool_calls: list[ToolCallResult],
        progress_callback: ProgressCallback | None,
    ) -> TurnResult:
        state.post_budget_submission_attempts += 1
        state.messages.append({"role": "assistant", "content": assistant_text})
        if state.post_budget_submission_attempts >= MAX_POST_BUDGET_SUBMISSION_ATTEMPTS:
            return TurnResult(done=True)
        _append_traced_message(
            state,
            self.adapter.format_tool_results(
                executed_tool_calls,
                nudge_suffix=_submission_required_nudge(state),
            ),
        )
        self._publish_progress(state, progress_callback)
        return TurnResult(done=False)

    def _handle_tool_results(
        self,
        state: ModelSessionState,
        *,
        assistant_text: str,
        executed_tool_calls: list[ToolCallResult],
        iteration_index: int,
        maximum_iterations: int,
        progress_callback: ProgressCallback | None,
    ) -> TurnResult:
        if state.done_signaled:
            return TurnResult(done=True)
        state.messages.append({"role": "assistant", "content": assistant_text})
        nudge = _nudge_suffix(self.adapter, state, iteration_index, maximum_iterations)
        if _is_terminal_budget_exhausted(state):
            nudge = _submission_required_nudge(state)
        _append_traced_message(
            state,
            self.adapter.format_tool_results(executed_tool_calls, nudge_suffix=nudge),
        )
        self._publish_progress(state, progress_callback)
        return TurnResult(done=False)

    def _publish_progress(
        self,
        state: ModelSessionState,
        progress_callback: ProgressCallback | None,
    ) -> None:
        if progress_callback is None:
            return
        progress_callback(
            self.build_agent_state(
                messages=state.messages,
                trajectory=state.trajectory,
            ),
            None,
        )


def _nudge_suffix(
    adapter: ModelAdapter,
    state: ModelSessionState,
    iteration_index: int,
    maximum_iterations: int,
) -> str:
    budget_suffix = _terminal_budget_suffix(state)
    nudge_threshold = int(maximum_iterations * adapter.investigation_nudge_fraction)
    if iteration_index != nudge_threshold - 1:
        return budget_suffix
    state.trajectory.append(
        TrajectoryEntry(
            entry_type="think",
            content="Running low on turns - wrapping up investigation.",
        )
    )
    turn_nudge = (
        "\n\nYou are running low on remaining turns. "
        "Finish your investigation and submit file-level results now."
    )
    return turn_nudge + budget_suffix


def _terminal_budget_suffix(state: ModelSessionState) -> str:
    if state.terminal_call_budget is None:
        return ""
    remaining = state.terminal_call_budget - state.terminal_calls_used
    return f"\n[{remaining} tool-calls remaining]"


def _submission_required_nudge(state: ModelSessionState) -> str:
    budget = state.terminal_call_budget
    return (
        f"\nERROR: Repository tool budget exhausted ({budget}/{budget}). "
        "Submit now using either "
        '<tool_call>{"name":"submit_vulnerable_files","arguments":{"ranked_files":'
        '["path/to/file"]}}</tool_call> or '
        '<tool_call>{"name":"submit_no_vulnerability_found","arguments":{}}</tool_call>.'
    )


def _is_terminal_budget_exhausted(state: ModelSessionState) -> bool:
    if state.terminal_call_budget is None:
        return False
    return state.terminal_calls_used >= state.terminal_call_budget


def _append_traced_message(
    state: ModelSessionState,
    message: dict[str, str],
) -> None:
    state.messages.append(message)
    state.session_trace.record_message(message)

File: vscode-extension/src/antares/agent/modelTurn.ts
// Port of antares_cli/agent/model_turn.py + model_turn_stream.py — one model turn.

import {
  InferenceBackend,
  InferenceContextLengthError,
  InferenceError,
  InferenceHttpError,
  InferenceStreamError,
} from "../inference/backend";
import { TrajectoryEntry } from "../output/finding";
import { ModelAdapter, ToolCallResult } from "./modelAdapter";
import { ModelSessionState, TurnResult } from "./state";
import { ParsedToolCall } from "./streamingParser";
import { SubmissionHandler } from "./submission";
import { AgentToolExecutor } from "./toolExecution";
import { AgentStateSnapshot, ProgressCallback } from "./types";

const MAX_PARSE_RETRIES = 2;
const MAX_MODEL_TURN_CHARS = 1_000_000;
const MAX_POST_BUDGET_SUBMISSION_ATTEMPTS = 3;

interface ParsedTurn {
  generationError: boolean;
  assistantText: string;
  toolCalls: ParsedToolCall[];
  doneSignaled: boolean;
  answerText: string | null;
  parseError: boolean;
}

interface ToolExecution {
  submitted: boolean;
  allDuplicates: boolean;
  executedToolCalls: ToolCallResult[];
}

export type BuildAgentStateFn = (
  messages: ModelSessionState["messages"],
  trajectory: TrajectoryEntry[]
) => AgentStateSnapshot;

export class ModelTurnRunner {
  constructor(
    private readonly backend: InferenceBackend,
    private readonly adapter: ModelAdapter,
    private readonly toolExecutor: AgentToolExecutor,
    private readonly submissionHandler: SubmissionHandler,
    private readonly buildAgentState: BuildAgentStateFn,
    private readonly isSubmitToolCall: (call: ParsedToolCall) => boolean
  ) {}

  async runTurn(
    state: ModelSessionState,
    iterationIndex: number,
    maximumIterations: number,
    progressCallback?: ProgressCallback
  ): Promise<TurnResult> {
    const parsed = await this.parseTurnWithRetries(state, progressCallback);
    if (parsed.generationError) {
      state.generationErrors += 1;
      state.reasoningLog.push(
        "Model generation error interrupted this run; results may be incomplete."
      );
      return { done: true };
    }

    if (isTerminalBudgetExhausted(state)) {
      const execution = this.executeToolCalls(parsed.toolCalls, state, progressCallback);
      if (execution.submitted || state.doneSignaled) {
        return { done: true };
      }
      return this.handlePostBudgetSubmissionAttempt(
        state,
        parsed.assistantText,
        execution.executedToolCalls,
        progressCallback
      );
    }

    if (parsed.toolCalls.length === 0) {
      return this.handleNoToolCalls(state, parsed.assistantText);
    }

    state.consecutiveNoToolTurns = 0;
    const execution = this.executeToolCalls(parsed.toolCalls, state, progressCallback);
    if (execution.submitted) {
      return { done: true };
    }
    if (execution.allDuplicates) {
      return this.handleDuplicateToolTurn(state, parsed.assistantText);
    }
    state.consecutiveDuplicateTurns = 0;
    return this.handleToolResults(
      state,
      parsed.assistantText,
      execution.executedToolCalls,
      iterationIndex,
      maximumIterations,
      progressCallback
    );
  }

  private async parseTurnWithRetries(
    state: ModelSessionState,
    progressCallback?: ProgressCallback
  ): Promise<ParsedTurn> {
    const retryParseErrors = !isTerminalBudgetExhausted(state);
    let attempt = await this.streamSingleAttempt(state, progressCallback);
    for (let retryIndex = 0; retryIndex < MAX_PARSE_RETRIES; retryIndex++) {
      if (attempt.generationError) {
        return attempt;
      }
      const shouldRetry =
        retryParseErrors &&
        attempt.parseError &&
        attempt.toolCalls.length === 0 &&
        !attempt.doneSignaled;
      if (!shouldRetry) {
        return attempt;
      }
      state.retriedTurnsCount += 1;
      state.reasoningLog.push(`Parse retry ${retryIndex + 1}/${MAX_PARSE_RETRIES}: retrying model turn`);
      attempt = await this.streamSingleAttempt(state, progressCallback);
    }
    return attempt;
  }

  private async streamSingleAttempt(
    state: ModelSessionState,
    progressCallback?: ProgressCallback
  ): Promise<ParsedTurn> {
    state.outputParser = this.adapter.createOutputParser();
    let assistantText = "";
    const toolCalls: ParsedToolCall[] = [];
    let doneSignaled = false;
    let answerText: string | null = null;
    let parseError = false;

    const consume = (events: ReturnType<typeof state.outputParser.feed>) => {
      for (const event of events) {
        if (event.kind === "tool_call") {
          toolCalls.push(event);
          this.publishProgress(state, progressCallback);
        } else if (event.kind === "done") {
          doneSignaled = true;
          state.trajectory.push({ entry_type: "think", content: "Investigation complete." });
        } else if (event.kind === "answer") {
          answerText = this.adapter.cleanModelText(event.text);
          doneSignaled = true;
        } else {
          parseError = parseError || event.text.startsWith("parse error:");
          const stripped = event.text.trim();
          if (stripped) {
            state.reasoningLog.push(stripped);
            state.trajectory.push({ entry_type: "think", content: stripped });
            this.publishProgress(state, progressCallback);
          }
        }
      }
    };

    try {
      for await (const chunk of this.backend.streamGenerate(state.messages)) {
        assistantText += chunk;
        if (assistantText.length > MAX_MODEL_TURN_CHARS) {
          return this.streamFailure(assistantText, toolCalls, doneSignaled, answerText);
        }
        consume(state.outputParser.feed(chunk));
      }
      consume(state.outputParser.flush());
    } catch (error) {
      if (error instanceof InferenceHttpError) {
        throw new InferenceError(httpErrorMessage(error));
      }
      if (error instanceof InferenceContextLengthError || error instanceof InferenceStreamError) {
        return this.streamFailure(assistantText, toolCalls, doneSignaled, answerText);
      }
      throw error;
    }

    return { generationError: false, assistantText, toolCalls, doneSignaled, answerText, parseError };
  }

  private streamFailure(
    assistantText: string,
    toolCalls: ParsedToolCall[],
    doneSignaled: boolean,
    answerText: string | null
  ): ParsedTurn {
    return { generationError: true, assistantText, toolCalls, doneSignaled, answerText, parseError: false };
  }

  private executeToolCalls(
    toolCalls: ParsedToolCall[],
    state: ModelSessionState,
    progressCallback?: ProgressCallback
  ): ToolExecution {
    const result: ToolExecution = { submitted: false, allDuplicates: true, executedToolCalls: [] };
    for (const parsedCall of toolCalls) {
      if (this.isSubmitToolCall(parsedCall)) {
        this.submissionHandler.handle(parsedCall, state, progressCallback);
        result.submitted = true;
        return result;
      }
      if (this.skipDuplicateToolCall(parsedCall, state)) {
        continue;
      }
      result.allDuplicates = false;
      result.executedToolCalls.push(this.toolExecutor.execute(parsedCall.toolName, parsedCall.arguments, state));
    }
    return result;
  }

  private skipDuplicateToolCall(parsedCall: ParsedToolCall, state: ModelSessionState): boolean {
    const dedupKey = stableStringify({ tool: parsedCall.toolName, args: parsedCall.arguments });
    if (!state.seenToolCalls.has(dedupKey)) {
      state.seenToolCalls.add(dedupKey);
      return false;
    }
    state.toolCallCount += 1;
    state.failedToolCallsCount += 1;
    state.reasoningLog.push(`Skipped duplicate tool call: ${parsedCall.toolName}`);
    return true;
  }

  private handleNoToolCalls(state: ModelSessionState, assistantText: string): TurnResult {
    const remainingText = this.adapter.cleanModelText(assistantText);
    if (remainingText && !state.answerText) {
      state.answerText = remainingText;
    }
    const retryIndex = state.consecutiveNoToolTurns;
    if (retryIndex < this.adapter.noToolRetryLimit && !state.doneSignaled) {
      state.consecutiveNoToolTurns += 1;
      state.messages.push({ role: "assistant", content: assistantText });
      appendTracedMessage(state, this.adapter.formatNoToolRetry(retryIndex));
      return { done: false };
    }
    return { done: true };
  }

  private handleDuplicateToolTurn(state: ModelSessionState, assistantText: string): TurnResult {
    state.consecutiveDuplicateTurns += 1;
    state.messages.push({ role: "assistant", content: assistantText });
    const forceSubmit = state.consecutiveDuplicateTurns >= 3;
    appendTracedMessage(state, this.adapter.formatDuplicateToolRetry(forceSubmit));
    if (forceSubmit) {
      state.reasoningLog.push("Loop stuck: 3 consecutive duplicate turns - forcing submit.");
      state.trajectory.push({
        entry_type: "think",
        content: "Stuck loop detected - forcing transition to submission.",
      });
    }
    return { done: false };
  }

  private handlePostBudgetSubmissionAttempt(
    state: ModelSessionState,
    assistantText: string,
    executedToolCalls: ToolCallResult[],
    progressCallback?: ProgressCallback
  ): TurnResult {
    state.postBudgetSubmissionAttempts += 1;
    state.messages.push({ role: "assistant", content: assistantText });
    if (state.postBudgetSubmissionAttempts >= MAX_POST_BUDGET_SUBMISSION_ATTEMPTS) {
      return { done: true };
    }
    appendTracedMessage(
      state,
      this.adapter.formatToolResults(executedToolCalls, submissionRequiredNudge(state))
    );
    this.publishProgress(state, progressCallback);
    return { done: false };
  }

  private handleToolResults(
    state: ModelSessionState,
    assistantText: string,
    executedToolCalls: ToolCallResult[],
    iterationIndex: number,
    maximumIterations: number,
    progressCallback?: ProgressCallback
  ): TurnResult {
    if (state.doneSignaled) {
      return { done: true };
    }
    state.messages.push({ role: "assistant", content: assistantText });
    let nudge = nudgeSuffix(this.adapter, state, iterationIndex, maximumIterations);
    if (isTerminalBudgetExhausted(state)) {
      nudge = submissionRequiredNudge(state);
    }
    appendTracedMessage(state, this.adapter.formatToolResults(executedToolCalls, nudge));
    this.publishProgress(state, progressCallback);
    return { done: false };
  }

  private publishProgress(state: ModelSessionState, progressCallback?: ProgressCallback): void {
    if (!progressCallback) {
      return;
    }
    progressCallback(this.buildAgentState(state.messages, state.trajectory), null);
  }
}

function isTerminalBudgetExhausted(state: ModelSessionState): boolean {
  if (state.terminalCallBudget === null) {
    return false;
  }
  return state.terminalCallsUsed >= state.terminalCallBudget;
}

function nudgeSuffix(
  adapter: ModelAdapter,
  state: ModelSessionState,
  iterationIndex: number,
  maximumIterations: number
): string {
  const budgetSuffix = terminalBudgetSuffix(state);
  const nudgeThreshold = Math.trunc(maximumIterations * adapter.investigationNudgeFraction);
  if (iterationIndex !== nudgeThreshold - 1) {
    return budgetSuffix;
  }
  state.trajectory.push({
    entry_type: "think",
    content: "Running low on turns - wrapping up investigation.",
  });
  return (
    "\n\nYou are running low on remaining turns. " +
    "Finish your investigation and submit file-level results now." +
    budgetSuffix
  );
}

function terminalBudgetSuffix(state: ModelSessionState): string {
  if (state.terminalCallBudget === null) {
    return "";
  }
  const remaining = state.terminalCallBudget - state.terminalCallsUsed;
  return `\n[${remaining} tool-calls remaining]`;
}

function submissionRequiredNudge(state: ModelSessionState): string {
  const budget = state.terminalCallBudget;
  return (
    `\nERROR: Repository tool budget exhausted (${budget}/${budget}). ` +
    "Submit now using either " +
    '<tool_call>{"name":"submit_vulnerable_files","arguments":{"ranked_files":' +
    '["path/to/file"]}}</tool_call> or ' +
    '<tool_call>{"name":"submit_no_vulnerability_found","arguments":{}}</tool_call>.'
  );
}

function appendTracedMessage(state: ModelSessionState, message: { role: string; content: string }): void {
  state.messages.push(message);
  state.sessionTrace.recordMessage(message);
}

function httpErrorMessage(error: InferenceHttpError): string {
  if (error.statusCode === 401) {
    return "Inference endpoint authentication failed (HTTP 401). Check the configured credentials.";
  }
  if (error.statusCode === 403) {
    return "Inference endpoint denied access (HTTP 403). Check the configured permissions.";
  }
  return `Inference endpoint request failed (HTTP ${error.statusCode}).`;
}

// Deterministic JSON with sorted keys (mirrors json.dumps(..., sort_keys=True)).
function stableStringify(value: unknown): string {
  return JSON.stringify(sortKeys(value));
}

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortKeys);
  }
  if (value && typeof value === "object") {
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      sorted[key] = sortKeys((value as Record<string, unknown>)[key]);
    }
    return sorted;
  }
  return value;
}
